Libplanet.RocksDBStore/RocksDBStore.cs

```diff
@@ -445,33 +530,56 @@ public override void PutBlock<T>(Block<T> block)
 
             byte[] key = BlockKey(block.Hash);
 
-            if (!(_blockDb.Get(key) is null))
+            if (!(_blockIndexDb.Get(key) is null))
             {
                 return;
             }
 
+            var timestamp = block.Timestamp.ToUnixTimeSeconds();
+
             foreach (Transaction<T> tx in block.Transactions)
             {
                 PutTransaction(tx);
             }
 
+            _rwBlockLock.EnterWriteLock();
+            string blockDbName = $"epoch{timestamp / _blockAtEachEpoch}";
+            RocksDBStoreEpoch epoch = new RocksDBStoreEpoch("Block", blockDbName);
+            if (!_blockDbCache.TryGetValue(epoch.DbName, out RocksDb blockDb))
+            {
+                blockDb = RocksDBUtils.OpenRocksDb(_options, BlockDbPath(epoch.DbName));
+                _blockDbCache.AddOrUpdate(epoch.DbName, blockDb);
+            }
+
             byte[] value = block.ToBlockDigest().Serialize();
-            _blockDb.Put(key, value);
+            blockDb.Put(key, value);
+            _blockIndexDb.Put(key, epoch.Serialize());
             _blockCache.AddOrUpdate(block.Hash, block.ToBlockDigest());
+            _rwBlockLock.ExitWriteLock();
```

Comment:
`_rwBlockLock.ExitWriteLock()` should be inside a `finally` block so that it's guaranteed to be called even if an any error happens.